research/fed-sm/job_configs/fedsm_prostate/app/custom/learners/supervised_monai_prostate_fedsm_learner.py

```diff
@@ -0,0 +1,392 @@
+        metric = self.local_valid(
+            self.fedsm_helper.person_model,
+            self.valid_loader,
+            abort_signal,
+            tb_id="val_metric_per_model",
+            current_round=current_round,
+        )
+        if abort_signal.triggered:
+            return make_reply(ReturnCode.TASK_ABORTED)
+        self.log_info(fl_ctx, f"val_metric_per_model: {metric:.4f}")
+        # save model
+        self.fedsm_helper.update_metric_save_person_model(current_round=current_round, metric=metric)
+
+        # compute delta models, initial models has the primary key set
+        local_weights = self.model.state_dict()
+        model_diff_global = {}
+        for name in global_weights:
+            if name not in local_weights:
+                continue
+            model_diff_global[name] = local_weights[name].cpu().numpy() - global_weights[name]
+            if np.any(np.isnan(model_diff_global[name])):
+                self.system_panic(f"{name} weights became NaN...", fl_ctx)
+                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        local_weights = self.fedsm_helper.person_model.state_dict()
+        model_diff_person = {}
+        for name in person_weights:
+            if name not in local_weights:
+                continue
+            model_diff_person[name] = local_weights[name].cpu().numpy() - person_weights[name]
+            if np.any(np.isnan(model_diff_person[name])):
+                self.system_panic(f"{name} weights became NaN...", fl_ctx)
+                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        local_weights = self.fedsm_helper.select_model.state_dict()
+        model_diff_select = {}
+        for name in select_weights:
+            if name not in local_weights:
+                continue
+            model_diff_select[name] = local_weights[name].cpu().numpy() - select_weights[name]
+            if np.any(np.isnan(model_diff_select[name])):
+                self.system_panic(f"{name} weights became NaN...", fl_ctx)
+                return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        optim_weights = self.fedsm_helper.select_optimizer.state_dict().get("state")
+        exp_avg = {}
+        exp_avg_sq = {}
+        for name in optim_weights:
+            exp_avg[str(name)] = optim_weights[name]["exp_avg"].cpu().numpy()
+            exp_avg_sq[str(name)] = optim_weights[name]["exp_avg_sq"].cpu().numpy()
+
+        # build the shareable
+        dxo_global_weights = DXO(data_kind=DataKind.WEIGHT_DIFF, data=model_diff_global)
+        dxo_person_weights = DXO(data_kind=DataKind.WEIGHT_DIFF, data=model_diff_person)
+        dxo_select_weights = DXO(data_kind=DataKind.WEIGHT_DIFF, data=model_diff_select)
+        dxo_select_exp_avg = DXO(data_kind=DataKind.WEIGHTS, data=exp_avg)
+        dxo_select_exp_avg_sq = DXO(data_kind=DataKind.WEIGHTS, data=exp_avg_sq)
+
+        dxo_dict = {
+            "global_weights": dxo_global_weights,
```

Comment:
The train function is very long. You could create the dictionary more directly, i.e. use `"global_weights": DXO(data_kind=DataKind.WEIGHT_DIFF, data=model_diff_global)` here.

Reply:
changed